polling_stations/apps/data_importers/base_importers.py

```diff
@@ -0,0 +1,874 @@
+                self.logger.log_message(
+                    logging.INFO,
+                    "address_record_to_dict() returned None with input:\n%s",
+                    variable=address,
+                    pretty=True,
+                )
+                continue
+
+            self.add_residential_address(address_info)
+
+    def add_residential_address(self, address_info):
+
+        if "council" not in address_info:
+            address_info["council"] = self.council
+
+        if "uprn" not in address_info:
+            address_info["uprn"] = ""
+        else:
+            # UPRNs less than 12 characters long may be left padded with zeros
+            # Making sure uprns in our addresslist are not left padded will help with matching them
+            # and catching duplicates.
+            address_info["uprn"] = str(address_info["uprn"]).lstrip("0")
+
+        self.addresses.append(address_info)
+
+
+class BaseStationsDistrictsImporter(BaseStationsImporter, BaseDistrictsImporter):
+    def pre_import(self):
+        raise NotImplementedError
+
+    @property
+    def districts_have_station_ids(self):
+
+        """
+        Check that we've called self.import_polling_{districts,stations}
+        Don't raise an exception though, because we might still want to
+        import just stations or districts for debugging - i.e. to see them
+        in qgis/the db. However if we don't also have the other half we
+        won't be able to update the UprnToCouncil table.
+        """
+        if len(self.districts.elements) < 1:
+            self.logger.log_message(
+                logging.WARNING, "No district records added to self.districts"
+            )
+        if len(self.stations.elements) < 1:
+            self.logger.log_message(
+                logging.WARNING, "No station records added to self.stations"
+            )
+
+        district_ids = {e.internal_council_id for e in self.districts.elements}
+        district_stations_ids = {e.polling_station_id for e in self.districts.elements}
+        station_ids = {e.internal_council_id for e in self.stations.elements}
+        station_districts_ids = {e.polling_district_id for e in self.stations.elements}
+
+        def check_for_mismatches(set_a, set_b):
+            if set_a - set_b:
+                self.logger.log_message(
+                    logging.WARNING,
+                    "No elements in %s have ids for following elements in %s: \n%s",
+                    variable=(set_b, set_a, set_a - set_b),
+                )
+
+        if district_stations_ids:
+            self.logger.log_message(logging.INFO, "Districts have station ids")
+            check_for_mismatches(district_stations_ids, station_ids)
+            return True
+
+        elif station_districts_ids:
+            self.logger.log_message(logging.INFO, "Stations have district ids")
+            check_for_mismatches(station_districts_ids, district_ids)
+            return False
```

Comment:
I haven't really deeply looked into this, but I think the reason why a whole load of the importers don't import any addresses (e.g: `import_southampton`, `import_salford`, `import_dundee2`, etc) is because these checks don't do what you're expecting. In python `bool({''}) == True` so I think you're always matching the first condition or something? Hopefully that gives you the jumping off point to fix them..

Reply:
Thanks, that was the issue for salford/dundee etc. I've pushed a fixup which catches that, has some better names and logging, and a test for it. 

I think Southampton is to do with new data on their end - there's a station element in the xml with no `point` tag - so just needs skipping in the `station_record_to_dict` method.
Changing line 64:
`geojson = record.geom.geojson`
to:
```python
if str(record["Name"]) == 'POLLING_STATIONS.147':
    return None
else:
    geojson = record.geom.geojson
```
Should fix it, and gives a sensible looking import - but probably only relevant for this pr as it's the only kml importer.